### PR TITLE
add support for subscribe / unsubscribe to a topic

### PR DESCRIPTION
**Pull request by chemidy**
work in progress, next week i have to work on documentation + manage topics list subscription for a device

### CODE REVIEW

File: fcm.go [hunk 1]
```diff
@@ -118,7 +120,64 @@ func (fcm *FCM) sendFirebaseRequest(payload Message) (*Response, error) {
...
+	// build response
+	response := new(Response)
+	if err := json.NewDecoder(resp.Body).Decode(response); err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (fcm *FCM) sendFirebaseTopicRequest(endpoint string, payload Topic) (*Response, error) {
```
Comment (acoshift) [inline fcm.go]: sendFirebaseRequest, sendFirebaseTopicRequest, SubscribeDeviceToTopic, etc. should accept context then add context to request using
`req = req.WithContext(ctx)`

File: fcm.go [hunk 2]
```diff
@@ -118,7 +120,64 @@ func (fcm *FCM) sendFirebaseRequest(payload Message) (*Response, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	// add headers
+	req.Header.Set("Authorization", fmt.Sprintf("key=%s", fcm.app.apiKey))
+	req.Header.Set("Content-Type", "application/json")
+
+	// TODO use fcm.app.client if possible (not working actually)
+	client := &http.Client{}
```
Comment (acoshift) [inline fcm.go]: you can move http.Client into FCM struct for reuse

File: fcm.go [hunk 3]
```diff
@@ -118,7 +120,64 @@ func (fcm *FCM) sendFirebaseRequest(payload Message) (*Response, error) {
...
+
+func (fcm *FCM) sendFirebaseTopicRequest(endpoint string, payload Topic) (*Response, error) {
+
+	// validate Topic
+	if err := payload.Validate(); err != nil {
+		return nil, err
+	}
+
+	// marshal Topic
+	data, err := json.Marshal(payload)
```
Comment (acoshift) [inline fcm.go]: you can create buffer and encode to it
```go
buf := bytes.Buffer{}
err := json.NewEncoder(&buf).Encode(payload)
// then use buf when create request
req, err := http.NewRequest(http.MethodPost, endpoint, &buf)
```

File: fcm.go [hunk 4]
```diff
@@ -118,7 +120,64 @@ func (fcm *FCM) sendFirebaseRequest(payload Message) (*Response, error) {
...
+	// TODO use fcm.app.client if possible (not working actually)
+	client := &http.Client{}
+
+	// execute request
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
```
Comment (acoshift) [inline fcm.go]: it is better to discard `resp.Body` when don't read the body
or just `defer io.Copy(ioutil.Discard, resp.Body)` after `defer resp.Body.Close()`

**Comment by acoshift** [comment]
@chemidy I can work on refactoring this if you don't have time.

File: fcm.go (hunk 2 above)
Comment (chemidy) [inline fcm.go]: Yes, it was planned on my list. It's done

File: fcm.go (hunk 1 above)
Comment (chemidy) [inline fcm.go]: ok, done thks for the review

File: fcm.go (hunk 3 above)
Comment (chemidy) [inline fcm.go]: ok done

File: fcm.go (hunk 4 above)
Comment (chemidy) [inline fcm.go]: ok thks done

**Comment by chemidy** [comment]
Thks for your review, i just update my branch.

**Pull request merged**